### PR TITLE
Add initial review of plugin

### PR DESCRIPTION
**Pull request by westonruter**
See commits.

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -111,15 +119,10 @@ function register_carousel_block_type() {
...
  */
 function render_carousel_block( $attributes, $content ) {
-	static $instance_id = 0;
-
 	if ( ! wp_script_is( 'gutenberg-bento-carousel-view' ) && ! is_admin() && ! is_amp() ) {
 		wp_enqueue_script( 'gutenberg-bento-carousel-view' );
 	}
 
-	$carousel_id = 'wp-block-gutenberg-bento-carousel-' . ++$instance_id;
-	$content     = str_replace( '<amp-base-carousel', "<amp-base-carousel id=\"$carousel_id\"", $content );
```
Comment (swissspidy) [inline inc/namespace.php:121]: Just note that an `id` attribute is needed for the amp-bind logic in the sanitizer. By default there is no `id`.

Comment (westonruter) [reply inc/namespace.php:121]: Yes. The `Document` abstraction in PHP has a `getElementId()` helper method that will return the `id` that is on the element, or else create one if it doesn't exist.

Comment (swissspidy) [reply inc/namespace.php:121]: Ohh, and I thought it's just a wrapper around `getAttribute('id')`. TIL!

Very useful 👍

File: block.json [hunk 2]
```diff
@@ -67,8 +67,8 @@
 	"supports": {
 		"align": true
 	},
-	"viewScript": "gutenberg-bento-carousel-view",
 	"editorScript": "gutenberg-bento-carousel-edit",
 	"editorStyle": "gutenberg-bento-carousel-edit",
+	"script": "gutenberg-bento-carousel-view",
```
Comment (westonruter) [inline block.json:72]: The `viewScript` setting seems to only be a Gutenberg thing and not yet in core, so it I switched it to `script`.

Comment (swissspidy) [reply block.json:72]: Yeah the `wp_enqueue_script` call in `render_carousel_block()` was basically a workaround for that until that lands in core.

Doesn't `script` cause this to run in the editor as well ([see docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#script))? This script is really only meant for the frontend.

File: inc/namespace.php [hunk 3]
```diff
@@ -28,32 +28,58 @@
...
-	wp_register_script( 'amp-runtime', 'https://cdn.ampproject.org/v0.js', array(), false );
+function register_bento_assets() {
+	if ( ! wp_script_is( AMP_RUNTIME_SCRIPT_HANDLE, 'registered' ) ) {
+		wp_register_script( AMP_RUNTIME_SCRIPT_HANDLE, 'https://cdn.ampproject.org/v0.js', array(), null );
+	}
 
-	wp_register_script( 'amp-bento-carousel', 'https://cdn.ampproject.org/v0/amp-base-carousel-1.0.js', array( 'amp-runtime' ), false );
+	// Note that amp-runtime will not eventually be a dependency for Bento.
+	$src    = sprintf( 'https://cdn.ampproject.org/v0/amp-base-carousel-%s.js', BASE_CAROUSEL_VERSION );
+	$script = wp_scripts()->query( Extension::BASE_CAROUSEL );
```
Comment (swissspidy) [inline inc/namespace.php:66]: This does not exist if the AMP plugin is not active.

Comment (westonruter) [reply inc/namespace.php:66]: Correct, so then it will be `false` and the `else` condition will register it.

File: block.json (hunk 2 above)
Comment (westonruter) [reply block.json:72]: Ah, good catch. I missed that.

File: inc/namespace.php (hunk 3 above)
Comment (swissspidy) [reply inc/namespace.php:70]: Sorry, I meant the Extension class itself. Not the script.

Comment (westonruter) [reply inc/namespace.php:70]: oh, haha. 🤦

Comment (westonruter) [reply inc/namespace.php]: Fixed in 5dbfe9c

File: inc/namespace.php [hunk 4]
```diff
@@ -101,25 +157,19 @@ function register_carousel_block_type() {
...
  */
 function render_carousel_block( $attributes, $content ) {
-	static $instance_id = 0;
-
-	if ( ! wp_script_is( 'gutenberg-bento-carousel-view' ) && ! is_admin() && ! is_amp() ) {
-		wp_enqueue_script( 'gutenberg-bento-carousel-view' );
+	// Note this is a temporary measure until Gutenberg 11 is live. See <https://github.com/WordPress/gutenberg/pull/32977>.
+	if ( ! is_admin() && ! is_amp() ) {
+		wp_enqueue_script( BLOCK_VIEW_SCRIPT_HANDLE );
 	}
```
Comment (swissspidy) [inline inc/namespace.php:171]: I just learned that `viewScript` will only be enqueued automatically if there is _no_ `render_callback`. This way one can conditionally enqueue the `viewScript` in the callback if needed.

Comment (swissspidy) [reply inc/namespace.php:171]: It's now in trunk btw: https://github.com/WordPress/wordpress-develop/pull/1412

File: inc/namespace.php [hunk 5]
```diff
@@ -26,62 +26,118 @@
...
+	// Note the is_amp() function can only be called at or after the `wp` action.
+	if ( ! is_amp() ) {
+		return;
+	}
+
+	// Prevent the view script from being enqueued on AMP pages, since it will be added automatically by the AMP plugin.
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( BLOCK_NAME );
+	if ( $block_type instanceof WP_Block_Type ) {
+		$block_type->script = null;
+	}
```
Comment (swissspidy) [inline inc/namespace.php:130]: Not needed anymore since you changed `script` back to `viewScript`

File: inc/namespace.php [hunk 6]
```diff
@@ -26,62 +26,118 @@
...
+
+	// Remove script/style dependencies which AMP will handle.
+	$view_script = wp_scripts()->query( BLOCK_VIEW_SCRIPT_HANDLE );
+	if ( $view_script ) {
+		$view_script->deps = array_diff( $view_script->deps, [ BASE_CAROUSEL_SCRIPT_HANDLE ] );
+	}
```
Comment (swissspidy) [inline inc/namespace.php:136]: `viewScript` will not be enqueued on AMP pages anyway.

**Review by swissspidy — APPROVED** [review]

**Comment by swissspidy** [comment]
Thanks a lot Weston! 🎉 🚀

**Comment by swissspidy** [comment]
OK not sure how that happened, but now in the AMP version the carousel is hidden 🤔 

According to dev tools it's because of this CSS:

https://github.com/ampproject/amphtml/blob/f916fb25531de0baa05d3ff7cd4feea4233e50ea/extensions/amp-base-carousel/1.0/amp-base-carousel.css#L28-L32

**Comment by westonruter** [comment]
I'm not seeing that on the AMP version. It's looking right to me:

image

Humm, actually try reloading the page a couple times. It may be this issue: https://github.com/ampproject/amphtml/issues/35485

To work around that, you could add this plugin temporarily:

```php
add_filter( 'amp_enable_ssr', '__return_false' );
```

**Pull request merged**